packages/yarnpkg-core/sources/Cache.ts

```diff
@@ -236,7 +236,40 @@ export class Cache {
     };
 
     const loadPackageThroughMutex = async () => {
-      const mutex = loadPackage();
+      const mutexedLoad = async () => {
+        // We don't yet know whether the cache path can be computed yet, since that
+        // depends on whether the cache is actually the mirror or not, and whether
+        // the checksum is known or not.
+        const tentativeCachePath = this.getLocatorPath(locator, expectedChecksum);
+
+        const cacheExists = tentativeCachePath !== null
+          ? await baseFs.existsPromise(tentativeCachePath)
+          : false;
+
+        const action = cacheExists
+          ? onHit
+          : onMiss;
+
+        // Note: must be synchronous, otherwise the mutex may break (a concurrent
+        // execution may start while we're running the action)
```

Comment:
That's not true anymore, since the mutex will be assigned in parallel of `mutexedLoad`. We can leave those callbacks sync, but better remove the comment.

Reply:
I knew I forgot to do something :smile:, was going to comment on that.